File: go/tests/feed_test.go
package relaytest

import (
	"net/http"
	"testing"

	"github.com/drim-dev/verifying-agent-code/go/harness"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/app"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/domain"
)

// ---- G-KAFKA producer: a post must await broker confirmation; broker down → 503, no row ----

func TestS_FD_01_broker_down_post_is_503(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	assertBrokerDownBlocksPost(t, fixture.BaseURL(), ada, ch)
}

func TestS_FD_01_naive_fire_and_forget_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	naive := fixture.NaiveApp(func(d *app.Deps) { d.Publisher = naiveKafkaPublisher{client: fixture.KafkaProducer()} })
	defer naive.Close()
	expectCatchToFail(t, "G-KAFKA", func(tb testing.TB) {
		assertBrokerDownBlocksPost(tb, naive.BaseURL(), ada, ch)
	})
}

func assertBrokerDownBlocksPost(tb testing.TB, baseURL string, sender domain.User, channelID string) {
	tb.Helper()
	ctx, cancel := bgCtx()
	defer cancel()
	if err := fixture.Kafka.StopBroker(ctx); err != nil {
		tb.Fatalf("stop broker: %v", err)
	}
	defer func() {
		startCtx, startCancel := bgCtx()
		defer startCancel()
		if err := fixture.Kafka.StartBroker(startCtx); err != nil {
			tb.Fatalf("restart broker: %v", err)
		}
	}()

	resp := clientAt(tb, baseURL, sender.ID).post("/channels/"+channelID+"/messages", map[string]any{"text": "while down"})
	resp.expect(http.StatusServiceUnavailable).expectCode("events:unavailable")
	if n := dbCountTB(tb, "channel_messages", "channel_id = $1", channelID); n != 0 {
		tb.Fatalf("expected no channel_messages row when the broker is down, got %d", n)
	}
}

func TestS_FD_02_fanout_to_members_except_sender(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	cleo := seedUser(t, "cleo")
	ch := seedChannel(t, ada, "team", false)
	seedJoin(t, bob, ch)
	seedJoin(t, cleo, ch)

	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "team news"}).expect(http.StatusCreated)
	awaitFeedSettled(t)

	for _, who := range []domain.User{bob, cleo} {
		entries := feedEntriesFor(t, who, ch)
		if len(entries) != 1 || entries[0] != "team news" {
			t.Fatalf("%s expected one feed entry, got %v", who.Handle, entries)
		}
	}
	if len(feedEntriesFor(t, ada, ch)) != 0 {
		t.Fatalf("sender should not get a feed entry")
	}
}

func TestS_FD_03_unread_increments(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	ch := seedChannel(t, ada, "team", false)
	seedJoin(t, bob, ch)

	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "one"}).expect(http.StatusCreated)
	awaitFeedSettled(t)
	if c := unreadFor(t, bob, ch); c != 1 {
		t.Fatalf("expected unread 1, got %d", c)
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "two"}).expect(http.StatusCreated)
	awaitFeedSettled(t)
	if c := unreadFor(t, bob, ch); c != 2 {
		t.Fatalf("expected unread 2, got %d", c)
	}
}

func TestS_FD_04_mark_read_resets_only_that_channel(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	a := seedChannel(t, ada, "alpha", false)
	b := seedChannel(t, ada, "beta", false)
	seedJoin(t, bob, a)
	seedJoin(t, bob, b)
	newClient(t, ada.ID).post("/channels/"+a+"/messages", map[string]any{"text": "a1"}).expect(http.StatusCreated)
	newClient(t, ada.ID).post("/channels/"+b+"/messages", map[string]any{"text": "b1"}).expect(http.StatusCreated)
	awaitFeedSettled(t)

	newClient(t, bob.ID).post("/channels/"+a+"/read", nil).expect(http.StatusNoContent)
	if c := unreadFor(t, bob, a); c != 0 {
		t.Fatalf("channel a unread should be 0, got %d", c)
	}
	if c := unreadFor(t, bob, b); c != 1 {
		t.Fatalf("channel b unread should be untouched at 1, got %d", c)
	}
}

// ---- G-KAFKA consumer: redelivery must not double-count feed/unread ----

func TestS_FD_05_redelivery_is_idempotent(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	ch := seedChannel(t, ada, "team", false)
	seedJoin(t, bob, ch)
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "once"}).expect(http.StatusCreated)
	awaitFeedSettled(t)

	ctx, cancel := bgCtx()
	defer cancel()
	ev := redeliveryEvent(t, ch, ada)

	// Correct consumer: re-publishing the same event leaves exactly one feed entry and unread 1.
	if err := fixture.Kafka.Publish(ctx, ev, harness.KafkaTopic); err != nil {
		t.Fatalf("republish: %v", err)
	}
	awaitFeedSettled(t)
	if got := len(feedEntriesFor(t, bob, ch)); got != 1 {
		t.Fatalf("expected one feed entry after redelivery, got %d", got)
	}
	if c := unreadFor(t, bob, ch); c != 1 {
		t.Fatalf("expected unread still 1 after redelivery, got %d", c)
	}
}

func TestS_FD_05_naive_unconditional_increment_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	ch := seedChannel(t, ada, "team", false)
	seedJoin(t, bob, ch)
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "once"}).expect(http.StatusCreated)
	awaitFeedSettled(t)

	ctx, cancel := bgCtx()
	defer cancel()
	naive := fixture.NaiveWorkers(ctx, func(d *app.Deps) {
		d.Feed = naiveFeedProjector{store: fixture.Store, unread: d.Unread, ids: d.IDs}
	})
	defer naive.Close()

	ev := redeliveryEvent(t, ch, ada)
	expectCatchToFail(t, "G-KAFKA", func(tb testing.TB) {
		// Two deliveries to the NAIVE topic; the naive projector increments unconditionally so
		// the unread counter runs ahead of the (deduplicated) feed.
		if err := fixture.Kafka.Publish(ctx, ev, harness.KafkaNaiveTopic); err != nil {
			tb.Fatalf("publish 1: %v", err)
		}
		if err := fixture.Kafka.AwaitConsumed(ctx, harness.KafkaNaiveTopic, harness.KafkaNaiveGroup); err != nil {
			tb.Fatalf("await 1: %v", err)
		}
		if err := fixture.Kafka.Publish(ctx, ev, harness.KafkaNaiveTopic); err != nil {
			tb.Fatalf("publish 2: %v", err)
		}
		if err := fixture.Kafka.AwaitConsumed(ctx, harness.KafkaNaiveTopic, harness.KafkaNaiveGroup); err != nil {
			tb.Fatalf("await 2: %v", err)
		}
		if c := unreadFor(tb, bob, ch); c != 1 {
			tb.Fatalf("naive double-count: unread should diverge from 1, got %d", c)
		}
	})
}

// ---- G-CACHE: a kicked member receives no further feed entries ----

func TestS_FD_06_kicked_member_gets_no_feed(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	bob := seedUser(t, "bob")
	ch := seedChannel(t, ada, "team", false)
	seedJoin(t, bob, ch)
	newClient(t, ada.ID).del("/channels/" + ch + "/members/" + bob.ID).expect(http.StatusNoContent)

	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "after kick"}).expect(http.StatusCreated)
	awaitFeedSettled(t)
	if got := len(feedEntriesFor(t, bob, ch)); got != 0 {
		t.Fatalf("kicked member should get no feed entry, got %d", got)
	}
	if c := unreadFor(t, bob, ch); c != 0 {
		t.Fatalf("kicked member unread should not increment, got %d", c)
	}
}

// ---- shared feed helpers ----

func awaitFeedSettled(t testing.TB) {
	t.Helper()
	ctx, cancel := bgCtx()
	defer cancel()
	if err := fixture.Kafka.AwaitConsumed(ctx, harness.KafkaTopic, harness.KafkaGroup); err != nil {
		t.Fatalf("await feed settled: %v", err)
	}
}

func feedEntriesFor(t testing.TB, who domain.User, channelID string) []string {
	t.Helper()
	var page struct {
		Items []struct {
			ChannelID string `json:"channelId"`
			Preview   string `json:"preview"`
		} `json:"items"`
	}
	clientAt(t, fixture.BaseURL(), who.ID).get("/feed").expect(http.StatusOK).decode(&page)
	previews := []string{}
	for _, it := range page.Items {
		if it.ChannelID == channelID {
			previews = append(previews, it.Preview)
		}
	}
	return previews
}

func unreadFor(t testing.TB, who domain.User, channelID string) int64 {
	t.Helper()
	var body struct {
		Channels map[string]int64 `json:"channels"`
	}
	clientAt(t, fixture.BaseURL(), who.ID).get("/me/unread").expect(http.StatusOK).decode(&body)
	return body.Channels[channelID]
}

// redeliveryEvent builds the message.posted event for the single message posted to channelID,
// so the test can republish it (exercising the consumer's idempotency directly).
func redeliveryEvent(t *testing.T, channelID string, sender domain.User) domain.MessagePosted {
	t.Helper()
	ctx, cancel := bgCtx()
	defer cancel()
	msgs, err := fixture.Store.ChannelMessages(ctx, channelID, "", 1)
	if err != nil || len(msgs) == 0 {
		t.Fatalf("load posted message: %v (n=%d)", err, len(msgs))
	}
	m := msgs[0]
	return domain.MessagePosted{
		MessageID: m.ID, ChannelID: channelID, SenderID: sender.ID,
		Preview: domain.Preview(m.Text), PostedAt: m.CreatedAt,
	}
}

File: python/tests/feed_test.py
"""Feed / unread / Kafka (S-FD) acceptance scenarios, including the catching
halves of the G-KAFKA (producer broker-down, consumer idempotency) and G-CACHE
(no feed for a kicked member) cases.
"""

from __future__ import annotations

from datetime import UTC, datetime

from harness.fixture import Fixture
from harness.kafka_harness import TOPIC
from relay import domain
from relay.infra.codecs import serialize_message_posted
from tests.helpers import (
    await_until,
    client_for,
    db_count,
    seed_channel,
    seed_member,
    seed_user,
)


def _channel_with_members(fixture: Fixture):
    a = seed_user(fixture, "fda")
    b = seed_user(fixture, "fdb")
    c = seed_user(fixture, "fdc")
    channel_id = seed_channel(fixture, a, "general", private=False)
    seed_member(fixture, a, channel_id, b)
    seed_member(fixture, a, channel_id, c)
    return a, b, c, channel_id


def test_s_fd_01_kafka_broker_down_is_503(fixture: Fixture) -> None:
    """[G-KAFKA] harness stops the broker; member posts → 503; no message row.
    (Broker restarted afterwards.)"""
    a, _b, _c, channel_id = _channel_with_members(fixture)
    fixture.kafka.stop_broker()
    try:
        response = client_for(fixture, a.id).post(
            f"/channels/{channel_id}/messages", {"text": "lost?"}
        )
        response.expect(503).expect_code("events:unavailable")
        assert db_count(fixture, "channel_messages", "channel_id = :c", {"c": channel_id}) == 0
    finally:
        fixture.kafka.start_broker()


def test_s_fd_02_fanout_to_members_except_sender(fixture: Fixture) -> None:
    """[G-KAFKA await-shape] A posts → feed entries for B and C with preview; none for A."""
    a, b, c, channel_id = _channel_with_members(fixture)
    client_for(fixture, a.id).post(
        f"/channels/{channel_id}/messages", {"text": "broadcast"}
    ).expect(201)

    await_until(lambda: db_count(fixture, "feed_entries", "user_id = :u", {"u": b.id}) == 1)
    await_until(lambda: db_count(fixture, "feed_entries", "user_id = :u", {"u": c.id}) == 1)
    assert db_count(fixture, "feed_entries", "user_id = :u", {"u": a.id}) == 0
    feed_b = client_for(fixture, b.id).get("/feed").json()["items"]
    assert feed_b[0]["preview"] == "broadcast"


def test_s_fd_03_unread_increments(fixture: Fixture) -> None:
    a, b, _c, channel_id = _channel_with_members(fixture)
    poster = client_for(fixture, a.id)
    poster.post(f"/channels/{channel_id}/messages", {"text": "one"}).expect(201)
    await_until(
        lambda: client_for(fixture, b.id).get("/me/unread").json()["channels"].get(channel_id) == 1
    )
    poster.post(f"/channels/{channel_id}/messages", {"text": "two"}).expect(201)
    await_until(
        lambda: client_for(fixture, b.id).get("/me/unread").json()["channels"].get(channel_id) == 2
    )


def test_s_fd_04_read_resets_counter(fixture: Fixture) -> None:
    a, b, _c, channel_id = _channel_with_members(fixture)
    other_channel = seed_channel(fixture, a, "other", private=False)
    seed_member(fixture, a, other_channel, b)

    client_for(fixture, a.id).post(f"/channels/{channel_id}/messages", {"text": "x"}).expect(201)
    client_for(fixture, a.id).post(
        f"/channels/{other_channel}/messages", {"text": "y"}
    ).expect(201)

    def unread_of(cid: str) -> int:
        return client_for(fixture, b.id).get("/me/unread").json()["channels"].get(cid, 0)

    await_until(lambda: unread_of(channel_id) == 1)
    await_until(lambda: unread_of(other_channel) == 1)

    client_for(fixture, b.id).post(f"/channels/{channel_id}/read").expect(204)
    counts = client_for(fixture, b.id).get("/me/unread").json()["channels"]
    assert counts.get(channel_id, 0) == 0
    assert counts.get(other_channel) == 1  # other channel untouched


def test_s_fd_05_consumer_idempotent_on_redelivery(fixture: Fixture) -> None:
    """[G-KAFKA] harness re-publishes the same event → still exactly one feed
    entry for B AND the unread counter still 1 (feed and counter must not diverge)."""
    a, b, _c, channel_id = _channel_with_members(fixture)
    response = client_for(fixture, a.id).post(
        f"/channels/{channel_id}/messages", {"text": "dup"}
    )
    response.expect(201)
    message_id = response.json()["id"]
    await_until(
        lambda: db_count(
            fixture,
            "feed_entries",
            "user_id = :u AND message_id = :m",
            {"u": b.id, "m": message_id},
        )
        == 1
    )

    # Re-publish the SAME event directly to the suite's topic.
    event = domain.MessagePosted(
        message_id=message_id,
        channel_id=channel_id,
        sender_id=a.id,
        preview="dup",
        posted_at=datetime.now(UTC),
    )
    fixture.kafka.publish(TOPIC, channel_id, serialize_message_posted(event))
    fixture.kafka.await_consumed(TOPIC, "feed-fanout")

    assert db_count(
        fixture, "feed_entries", "user_id = :u AND message_id = :m", {"u": b.id, "m": message_id}
    ) == 1
    assert client_for(fixture, b.id).get("/me/unread").json()["channels"].get(channel_id) == 1


def test_s_fd_06_cache_no_feed_for_kicked_member(fixture: Fixture) -> None:
    """[G-CACHE] owner kicks B, then A posts → no new feed entry for B; B's unread
    counter for the channel not incremented."""
    a, b, _c, channel_id = _channel_with_members(fixture)
    client_for(fixture, a.id).delete(f"/channels/{channel_id}/members/{b.id}").expect(204)

    response = client_for(fixture, a.id).post(
        f"/channels/{channel_id}/messages", {"text": "after-kick"}
    )
    response.expect(201)
    message_id = response.json()["id"]
    fixture.kafka.await_consumed(TOPIC, "feed-fanout")

    assert db_count(
        fixture, "feed_entries", "user_id = :u AND message_id = :m", {"u": b.id, "m": message_id}
    ) == 0
    assert client_for(fixture, b.id).get("/me/unread").json()["channels"].get(channel_id, 0) == 0
